internal/cms/signed.go

```diff
@@ -0,0 +1,324 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cms
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/x509"
+	"encoding/asn1"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/notaryproject/tspclient-go/internal/encoding/asn1/ber"
+	"github.com/notaryproject/tspclient-go/internal/hashutil"
+	"github.com/notaryproject/tspclient-go/internal/oid"
+)
+
+// ParsedSignedData is a parsed SignedData structure for golang friendly types.
+type ParsedSignedData struct {
+	// Content is the content of the EncapsulatedContentInfo.
+	Content []byte
+
+	// ContentType is the content type of the EncapsulatedContentInfo.
+	ContentType asn1.ObjectIdentifier
+
+	// Certificates is the list of certificates in the SignedData.
+	Certificates []*x509.Certificate
+
+	// CRLs is the list of certificate revocation lists in the SignedData.
+	CRLs []x509.RevocationList
+
+	// SignerInfos is the list of signer information in the SignedData.
+	SignerInfos []SignerInfo
+}
+
+// ParseSignedData parses ASN.1 BER-encoded SignedData structure to golang
+// friendly types.
+//
+// Only supported SignedData version is 3.
+func ParseSignedData(berData []byte) (*ParsedSignedData, error) {
+	data, err := ber.ConvertToDER(berData)
+	if err != nil {
+		return nil, SyntaxError{Message: "invalid signed data: failed to convert from BER to DER", Detail: err}
+	}
+	var contentInfo ContentInfo
+	if _, err := asn1.Unmarshal(data, &contentInfo); err != nil {
+		return nil, SyntaxError{Message: "invalid content info: failed to unmarshal DER to ContentInfo", Detail: err}
+	}
+	if !oid.SignedData.Equal(contentInfo.ContentType) {
+		return nil, ErrNotSignedData
+	}
+
+	var signedData SignedData
+	if _, err := asn1.Unmarshal(contentInfo.Content.Bytes, &signedData); err != nil {
+		return nil, SyntaxError{Message: "invalid signed data", Detail: err}
+	}
+
+	if signedData.Version != 3 {
+		return nil, SyntaxError{Message: fmt.Sprintf("unsupported signed data version: got %d, want 3", signedData.Version)}
+	}
+
+	certs, err := x509.ParseCertificates(signedData.Certificates.Bytes)
```

Comment:
We should validate the certificate meets timestamp certificate requirement as per CA/B forum.
https://github.com/cabforum/code-signing/blob/v3.7/docs/CSBR.md

This should be performed as part of parse operation because caller of the function would expect the library to do basic checks.